File: app/src/test/java/com/gytxtx/openjbd/protocol/JbdFrameAssemblerTest.java
package com.gytxtx.openjbd.protocol;

import org.junit.Test;

import java.util.Arrays;
import java.util.List;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertTrue;

public final class JbdFrameAssemblerTest {
    @Test
    public void append_reassemblesSplitFrame() {
        JbdFrameAssembler assembler = new JbdFrameAssembler();
        byte[] frame = JbdFrameTest.responseFrame(JbdCommands.CMD_BASIC_INFO, 0, new byte[]{0x01, 0x02});

        assertTrue(assembler.append(Arrays.copyOfRange(frame, 0, 4)).isEmpty());
        List<JbdFrame> frames = assembler.append(Arrays.copyOfRange(frame, 4, frame.length));

        assertEquals(1, frames.size());
        assertEquals(JbdCommands.CMD_BASIC_INFO & 0xFF, frames.get(0).command);
    }

    @Test
    public void append_extractsMultipleFramesAndSkipsNoise() {
        JbdFrameAssembler assembler = new JbdFrameAssembler();
        byte[] first = JbdFrameTest.responseFrame(JbdCommands.CMD_BASIC_INFO, 0, new byte[]{0x01});
        byte[] second = JbdFrameTest.responseFrame(JbdCommands.CMD_CELL_VOLTAGE, 0, new byte[]{0x0C, (byte) 0xE4});
        byte[] data = concat(new byte[]{0x55, 0x66}, first, second);

        List<JbdFrame> frames = assembler.append(data);

        assertEquals(2, frames.size());
        assertEquals(JbdCommands.CMD_BASIC_INFO & 0xFF, frames.get(0).command);
        assertEquals(JbdCommands.CMD_CELL_VOLTAGE & 0xFF, frames.get(1).command);
    }

    private static byte[] concat(byte[]... chunks) {
        int size = 0;
        for (byte[] chunk : chunks) {
            size += chunk.length;
        }
        byte[] out = new byte[size];
        int offset = 0;
        for (byte[] chunk : chunks) {
            System.arraycopy(chunk, 0, out, offset, chunk.length);
            offset += chunk.length;
        }
        return out;
    }
}

File: app/src/test/kotlin/com/gytxtx/openjbd/protocol/JbdFrameAssemblerTest.kt
package com.gytxtx.openjbd.protocol

import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Test

class JbdFrameAssemblerTest {
    @Test
    fun append_reassemblesSplitFrame() {
        val assembler = JbdFrameAssembler()
        val frame = JbdFrameTest.responseFrame(JbdCommands.CMD_BASIC_INFO, 0, byteArrayOf(0x01, 0x02))

        assertTrue(assembler.append(frame.copyOfRange(0, 4)).isEmpty())
        val frames = assembler.append(frame.copyOfRange(4, frame.size))

        assertEquals(1, frames.size)
        assertEquals(JbdCommands.CMD_BASIC_INFO.toInt() and 0xFF, frames[0].command)
    }

    @Test
    fun append_extractsMultipleFramesAndSkipsNoise() {
        val assembler = JbdFrameAssembler()
        val first = JbdFrameTest.responseFrame(JbdCommands.CMD_BASIC_INFO, 0, byteArrayOf(0x01))
        val second = JbdFrameTest.responseFrame(JbdCommands.CMD_CELL_VOLTAGE, 0, byteArrayOf(0x0C, 0xE4.toByte()))
        val data = byteArrayOf(0x55, 0x66) + first + second

        val frames = assembler.append(data)

        assertEquals(2, frames.size)
        assertEquals(JbdCommands.CMD_BASIC_INFO.toInt() and 0xFF, frames[0].command)
        assertEquals(JbdCommands.CMD_CELL_VOLTAGE.toInt() and 0xFF, frames[1].command)
    }
}
